### ISSUE TITLE
False error with control flow type analysis

### ISSUE THREAD

**Issue by zakjan** [Microsoft/TypeScript#17243]
**TypeScript Version:**  2.4.0

**Code**

```ts
interface Item {
    a: {
        b?: {
            c: {
                d?: string[];
            };
        }[];
    };
}

function test(item: Item): string | undefined {
    return item.a.b && item.a.b[0] && item.a.b[0].c.d && item.a.b[0].c.d[0];
}

test({a: {}});
```

**Expected behavior:**
no error is thrown

**Actual behavior:**
`error TS2532: Object is possibly 'undefined'.` at `_item.a.b[0].c.d_[0]`

**Comment by andy-ms** [comment]
A simpler example:
```ts
function f(foo: Array<{ x: number } | undefined>) {
    foo[0] && foo[0].x;
}
```

I don't think we support this for computed properties (including array entries). You could make a local variable for `foo[0]` to fix it.

**Comment by zakjan** [comment]
How come it doesn't throw the same error at `_item.a.b_[0]`?

> You could make a local variable for arr[0] to fix it.
Yes, but with such deeply nested structures it becomes very annoying.

**Comment by andy-ms** [comment]
You mean in `item.a.b && item.a.b[0]`? You just tested `item.a.b`, so TS knows that's defined.
You may also be interested in https://github.com/tc39/proposal-optional-chaining .

**Comment by RyanCavanaugh** [comment]
See #10530 - type guards currently do not work on element access expressions

**Comment by zakjan** [comment]
> You mean in `item.a.b && item.a.b[0]`? You just tested `item.a.b`, so TS knows that's defined.
Yes, similarly I tested `item.a.b && item.a.b[0] && item.a.b[0].c.d && item.a.b[0].c.d`, so TS could know it is defined.

**Comment by mhegazy** [comment]
Automatically closing this issue for housekeeping purposes. The issue labels indicate that it is unactionable at the moment or has already been addressed.
